generic/tests/trans_hugepage_defrag.py

```diff
@@ -88,74 +58,46 @@ def change_feature_status(test, status, feature_path, test_config):
-        try:
-            test.log.info("Prepare tmpfs in host")
-            if not os.path.isdir(mem_path):
-                os.makedirs(mem_path)
-            process.run("mount -t tmpfs none %s" % mem_path, shell=True)
-            test.log.info("Start using dd to fragment memory in guest")
-            cmd = ("for i in `seq 262144`; do dd if=/dev/urandom of=%s/$i "
-                   "bs=4K count=1 & done" % mem_path)
-            process.run(cmd, shell=True)
-        finally:
-            process.run("umount %s" % mem_path, shell=True)
-
-    test_config = test_setup.TransparentHugePageConfig(test, params)
+    tmp_directory = params.get("tmp_directory", "/var/tmp")
+    test_bin = params.get("test_bin", "thp_fragment")
+    source_file = params.get("source_file", "thp_fragment.c")
+    source_package = params.get("source_package", "thp_fragment.tar.gz")
+    host_path = utils_misc.get_path(data_dir.get_deps_dir('thp_defrag_tool'),
+                                    source_package)
+    copy_cmd = "cp -rf %s %s" % (host_path, tmp_directory)
+    process.system(copy_cmd, ignore_status=True, shell=True)
+    extract_cmd = "cd %s; tar xzvf %s" % (tmp_directory, source_package)
+    process.system(extract_cmd, ignore_status=True, shell=True)
+    build_cmd = "cd %s; gcc -lrt %s -o %s" % (tmp_directory,
+                                              source_file,
+                                              test_bin)
+    error_context.context("Build binary file '%s'" % test_bin, test.log.info)
+    process.run(build_cmd, ignore_status=True, shell=True)
+    test_config = test_setup.TransparentHugePageConfig(test, params, env)
     test.log.info("Defrag test start")
-    login_timeout = float(params.get("login_timeout", 360))
-    mem_path = os.path.join("/tmp", "thp_space")
-
-    try:
-        error_context.context("deactivating khugepaged defrag functionality")
-        change_feature_status(test, "off", "khugepaged/defrag", test_config)
-        change_feature_status(test, "off", "defrag", test_config)
-
-        vm = env.get_vm(params.get("main_vm"))
-        session = vm.wait_for_login(timeout=login_timeout)
-
-        fragment_host_memory(mem_path)
-
-        total = get_mem_stat('MemTotal')
-        hugepagesize = get_mem_stat('Hugepagesize')
-        nr_full = int(0.8 * (total / hugepagesize))
 
-        nr_hp_before = set_libhugetlbfs(nr_full)
+    error_context.context("deactivating khugepaged defrag functionality",
+                          test.log.info)
+    change_feature_status("off", "khugepaged/defrag", test_config)
+    change_feature_status("off", "defrag", test_config)
 
-        error_context.context("activating khugepaged defrag functionality")
-        change_feature_status(test, "on", "khugepaged/defrag", test_config)
-        change_feature_status(test, "on", "defrag", test_config)
+    thps_defrag_off = int(process.getoutput("/var/tmp/thp_fragment", shell=True).split()[1])
```

Comment:
It is the full path of the test_bin,  suggest using parameter, refer to the comment for L23 in cfg.

Reply:
Updated, thanks !